File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        crs_preq_map = {i : [] for i in range(numCourses)}
        for crs, prereq in prerequisites:
            crs_preq_map[crs].append(prereq)

        # ** Topological Sort ** #
        
        visited = [False] * numCourses
        cycle = [False] * numCourses
        # def dfs(crs):
        #     if crs in visited:
        #         return False
        #     if not crs_preq_map[crs]: return True
        #     visited.add(crs)
        #     for prereq in crs_preq_map[crs]:
        #         if not dfs(prereq): return False
        #     visited.remove(crs)
        #     crs_preq_map[crs] = []
        #     return True

        def dfs(crs):
            visited[crs] = True
            cycle[crs] = True
            for neigh in crs_preq_map[crs]:
                if not visited[neigh]:
                    if not dfs(neigh):
                        return False
                elif cycle[neigh]:
                    return False
            
            cycle[crs] = False
            return True

        for crs in crs_preq_map:
            if not dfs(crs): return False
            
        return True

File: 0207-course-schedule/0207-course-schedule.go
func canFinish(numCourses int, prerequisites [][]int) bool {
    adj_list := make(map[int][]int, numCourses)

    for _, pair := range prerequisites {
        u, v := pair[1], pair[0]

        adj_list[u] = append(adj_list[u], v)
    }

    inDegree := make([]int, numCourses)

    for _, v := range adj_list {
        for _, crs := range v {
            inDegree[crs]++
        }
    }

    
    queue := make([]int, 0)

    for crs, v := range inDegree {
        if v == 0 {
            queue = append(queue, crs)
        }
    }

    topoCount := 0
    for len(queue) > 0 {
        size := len(queue)
        for i := 0; i < size; i++ {
            cur := queue[0]
            queue = queue[1:]
            topoCount++

            for _, pre := range adj_list[cur] {
                inDegree[pre]--

                if inDegree[pre] == 0 {
                    queue = append(queue, pre)
                }
            }
        }
    }

    return topoCount == numCourses

}